IPM fix for esp32s3 targets

marekmatej opened the pull request:
This PR fixes how the APPCPU is being built in IPM samples. C array encapsulation is removed and instead normal build procedure is used. Fixing the `sysbuild` builds as well.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/80890

marekmatej committed:
dts: esp32s3: shm nodes update

Align the shared memories with the memory.h layout.
Reorder nodes to show memory related nodes together.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej committed:
soc: esp32s3: AMP support

Updates and fixes to support APPCPU.
- fix ld scripts
- fix and update memory layout
- fix build issues
- fix sysbuild

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej committed:
samples: drivers: IPM fixes

Updates and fixes to support IPM sample on ESP32:

- fix IPM sample code for APPCPU and PROCPU
- align with memory layout, add flash awarenes
- shell commands to stop/start APPCPU
- reorganize overlays

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej committed:
drivers: clock_control: amp clock fix

Avoid APPCPU to interact with a clock settings.
Fix warning when LOG_LEVEL_DBG.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

uLipe commented on soc/espressif/esp32s3/Kconfig.amp:
```diff
@@ -0,0 +1,31 @@
+config ESP32S3_APPCPU_IRAM_SIZE
+	hex "ESP32S3 APPCPU IRAM size"
+	depends on SOC_ESP32S3_PROCPU || SOC_ESP32S3_APPCPU
+	default 0x10000
```
why it was decreased to 64KB?

marekmatej replied:
It is the same setup as before due to a different approach. Now both IRAM and DRAM adds to allocate the space.

uLipe commented on soc/espressif/esp32s3/Kconfig.amp:
```diff
@@ -0,0 +1,31 @@
+	  Defines APPCPU DRAM area size in bytes.
+
+config SOC_ENABLE_APPCPU
+	bool "Utilize CPU1 of ESP32S3"
```
If this config is hidden you can remove the option prompt, otherwise it will became user selectable.

uLipe commented on soc/espressif/esp32s3/esp32s3-mp.c:
```diff
@@ -35,7 +38,139 @@ void esp_appcpu_start(void *entry_point)
+void esp_appcpu_image_stop(void)
+{
+#if 0
+	/* TODO: */
```
Is it still a Todo?

marekmatej replied:
missed this

uLipe commented on soc/espressif/esp32s3/esp32s3-mp.c:
```diff
@@ -35,7 +38,139 @@ void esp_appcpu_start(void *entry_point)
+	unsigned int entry_addr = 0;
+
+	if (started) {
+		printk("APPCPU allready started.\r\n");
```
Typo: allready

uLipe commented on soc/espressif/esp32s3/esp32s3-mp.c:
```diff
@@ -35,7 +38,139 @@ void esp_appcpu_start(void *entry_point)
+
+int esp_appcpu_init(void)
+{
+	esp_appcpu_image_start(0x20);
```
What this 0x20 mean?

marekmatej replied:
added explanation

uLipe reviewed (commented):
Great to see this being sent.

I just left some initial commits.

marekmatej commented on samples/drivers/ipm/ipm_esp32/src/main.c:
```diff
@@ -38,15 +41,23 @@ int main(void)
+		}
 
-		k_sleep(K_MSEC(200));
+		k_sleep(K_MSEC(1000));
```
Just a user experience.

marekmatej commented on dts/xtensa/espressif/esp32s3/esp32s3_common.dtsi:
```diff
@@ -91,14 +91,39 @@
+			reg = <0x3fcb2400 0x3c00>;
+		};
+
+		ipm0: ipm@3fcb6000 {
```
same, reflects the updated layout from `memory.h`

marekmatej commented on soc/espressif/esp32s3/Kconfig:
```diff
@@ -14,27 +14,17 @@ config SOC_SERIES_ESP32S3
-	hex "ESP32S3 APPCPU IRAM size"
-	depends on SOC_ESP32S3_PROCPU || SOC_ESP32S3_APPCPU
-	default 0x20000
+config ESP32_PHY_MAX_WIFI_TX_POWER
```
fixed

marekmatej committed:
boards: esp32s3: AMP support and fixes

Fix missing flash and code partition.
Add missing dts entries and use common partition tables to all related
non-Espressif boards, previously ommited.
Add uart1 node in pinctrl for APPCPU.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej committed:
manifest: update hal_espressif

Fetch latest features to support fixes.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

uLipe commented on soc/espressif/esp32s3/esp32s3-mp.c:
```diff
@@ -35,7 +38,135 @@ void esp_appcpu_start(void *entry_point)
+
+	for (int i = 0; i < src_len / 4; i++) {
+		dst[i] = data[i];
+	}
```
Just to avoid problems with cache coherence, it would not be good to invalidate D-Cache here before each time we move the data? Of course just in case we're moving data from the flash to the DRAM.

marekmatej replied:
Since updating ld script for the APPCPU did not get the `STORE_PROHIBITED` exception anymore, and I'd rather not change this now.

uLipe commented on soc/espressif/esp32s3/esp32s3-mp.c:
```diff
@@ -35,7 +38,135 @@ void esp_appcpu_start(void *entry_point)
+	const uint32_t fa_size = FIXED_PARTITION_SIZE(slot0_appcpu_partition);
+	const uint8_t fa_id = FIXED_PARTITION_ID(slot0_appcpu_partition);
+	int rc = 0;
+
```
Maybe same here, we can do even better, before loading the image, disable both I and D cache to avoid cache coherency issues.

uLipe commented on soc/espressif/esp32s3/esp32s3-mp.c:
```diff
@@ -35,7 +38,135 @@ void esp_appcpu_start(void *entry_point)
+
+void esp_appcpu_image_stop(void)
+{
+	esp_cpu_stall(1);
```
What about replace 1 to ESP32_APPCPU_INDEX or something in that sense?

marekmatej replied:
we have 2 main cores 0 and 1, don't think this is necessary.

uLipe commented on soc/espressif/esp32s3/soc.c:
```diff
@@ -213,3 +173,8 @@ void sys_arch_reboot(int type)
+
+#if defined(CONFIG_SOC_ENABLE_APPCPU) && !defined(CONFIG_MCUBOOT)
+extern int esp_appcpu_init(void);
+SYS_INIT(esp_appcpu_init, POST_KERNEL, 99);
```
Why not use CONFIG_APPLICATION_PRIORITY instead of the plain 99?

uLipe reviewed (commented):
I think we're almost there, I just left some extra questions

uLipe commented:
@mmahadevan108 any chance to put this into the 4.0? 

It is an important fix that will enable some samples for esp32 that are broken since HVM2 introduction.

Thank you :)

mmahadevan108 commented:
pinging @nordicjm for a review from HWV2 side as these samples were broken since the switch to HWV2.

mmahadevan108 commented:
@uLipe . This is a fairly big change. Is everything required for fixing the bug?
It might be easier to pull this into 4.0 if we can focus this PR to only include the changes needed to fix the bug and move the remaining to a separate PR that can be merged post 4.0 release.

marekmatej commented:
> This is a fairly big change.

@mmahadevan108 you are right, and the reason it is so hefty is it not only fixes the IPM build itself but also stuff related to HWMv2, eg using sysbuild for multi-image builds.

mmahadevan108 commented:
@uLipe @marekmatej thank you for clarifying. This PR needs approval from @kartben as he is the assignee before we can merge

marekmatej commented:
Now that assignee has been updated, I guess we can proceed as is? @mmahadevan108

uLipe commented:
I'd say the same thig
